flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java

```diff
@@ -0,0 +1,365 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperatorV2;
+import org.apache.flink.streaming.api.operators.BoundedMultiInput;
+import org.apache.flink.streaming.api.operators.Input;
+import org.apache.flink.streaming.api.operators.InputSelectable;
+import org.apache.flink.streaming.api.operators.InputSelection;
+import org.apache.flink.streaming.api.operators.MultipleInputStreamOperator;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+import org.apache.flink.streaming.api.operators.StreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.StreamOperatorParameters;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.util.TestAnyModeMultipleInputStreamOperator;
+import org.apache.flink.streaming.util.TestAnyModeMultipleInputStreamOperator.ToStringInput;
+import org.apache.flink.streaming.util.TestSequentialMultipleInputStreamOperator;
+import org.apache.flink.util.ExceptionUtils;
+
+import org.junit.Test;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Queue;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Test selective reading.
+ */
+public class StreamTaskMultipleInputSelectiveReadingTest {
+
+	@Test
+	public void testAnyOrderedReading() throws Exception {
+		ArrayDeque<Object> expectedOutput = new ArrayDeque<>();
+		expectedOutput.add(new StreamRecord<>("[1]: Hello-1"));
+		expectedOutput.add(new StreamRecord<>("[2]: 1"));
+		expectedOutput.add(new StreamRecord<>("[1]: Hello-2"));
+		expectedOutput.add(new StreamRecord<>("[2]: 2"));
+		expectedOutput.add(new StreamRecord<>("[1]: Hello-3"));
+		expectedOutput.add(new StreamRecord<>("[2]: 3"));
+		expectedOutput.add(new StreamRecord<>("[2]: 4"));
+
+		testBase(new TestAnyModeMultipleInputStreamOperator.Factory(), false, expectedOutput, true);
```

Comment:
It's again not easy to see what all the operators are doing. Would it be possible to have a generic operator, where you pass a `InputSelectable`, such that it is defined at the same place as the expected output?

Reply:
Hmmm, I think that would be quite difficult to abstract. I get your point, but I'm not sure if it's worth it. (especially take a look at the logic inside `SpecialRuleReadingStreamOperatorFactory`).

Reply:
Yes the logic would be hard to generalize. It may be worth just to do it partially for the easy cases.

Reply:
But `TestAnyModeMultipleInputStreamOperator` is the only easy case that can be easily generalised. Unifying it with `TestSequentialMultipleInputStreamOperator` would basically require to pass the whole operator here, via the generalised factory. I really don't think it makes sense And generalising just `TestAnyModeMultipleInputStreamOperator` also doesn't make sense, as that would just add code complexity?

Reply:
Okay keep as is.